Build improvements with UniFFI library mode [ci full]

bendk opened the pull request:
Added a tool to run uniffi-bindgen in library mode.  It can input either a specific library path or the megazord crate name.

Use that simplify several build scripts -- especially the generate docs ones.  The best part of this is that we no longer have to maintain hand-written modulemaps, which makes adding a new component harder than it needs to be.

Added a hack to do something like https://github.com/mozilla/uniffi-rs/pull/2235, which won't be released upstream until 0.29.  I wanted to get this functionality in without working through a UniFFI breaking change.

One change is that automation/swift-components-docs/generate-swift-project.sh now uses `megazord_ios` rather than `megazord`.  I think this should result in slightly more accurate docs, since historically some components in the Android megazord aren't in the iOS one (Although, I think they match at the present).



Do not merge until we get a UniFFI release with the functionality that we need, putting this up for early review until then though.

### Pull Request checklist ###
<!-- Before submitting the PR, please address each item -->
- **Breaking changes**:  This PR follows our [breaking change policy](https://github.com/mozilla/application-services/blob/main/docs/howtos/breaking-changes.md)
  - [ ] This PR follows the breaking change policy:
     - This PR has no breaking API changes, or
     - There are corresponding PRs for our consumer applications that resolve the breaking changes and have been approved
- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGELOG.md](../CHANGELOG.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due diligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

bendk commented:
I added `[ci full]` and will test a local build using firefox-android.  Can someone test a local firefox-ios build against this code?

bendk commented:
This is still not ready to merge until we get upstream changes in UniFFI, but I think the app-services code is ready for review.  @gruberb can you take a look and also check if a local iOs build works with these changes?  I tested Android and it worked for me.

gruberb commented on automation/swift-components-docs/generate-swift-project.sh:
```diff
@@ -17,20 +16,9 @@ INCLUDE_DIR="$WORKING_DIR/Sources/$FRAMEWORK_NAME/include"
-$CARGO uniffi-bindgen generate --library "$LIBRARY_FILE" --language swift --out-dir "$SWIFT_DIR"
+# Generate Swift bindings, headers, and module map to use as the documentation source
+echo "Generating Swift bindings"
+$CARGO uniffi-bindgen-library-mode megazord_ios swift "$SWIFT_DIR"
```
Forgot the `-m` parameter
```bash
$CARGO uniffi-bindgen-library-mode -m megazord_ios swift "$SWIFT_DIR"
```

gruberb reviewed (dismissed):
Building `Swift` docs fails because some typos and a missing parameter. But once fixed, it works 👍

bendk commented on megazords/ios-rust/build-xcframework.sh:
```diff
@@ -147,14 +146,15 @@ UNIFFI_BINDGEN_LIBRARY="$TARGET_DIR/aarch64-apple-ios/$BUILD_PROFILE/$LIB_NAME"
 # now we generate both the `.h` and `.swift` files, then delete the `.swift`.  Bleh.
 
-$CARGO uniffi-bindgen generate --library "$UNIFFI_BINDGEN_LIBRARY" -l swift -o "$COMMON/Headers"
+"$CARGO" uniffi-bindgen-library-mode -l "$UNIFFI_BINDGEN_LIBRARY" swift "$COMMON/Headers" --framework-modulemap
```
I tried, but it ended up being to tricky because this script cross-compiles the megazord.  I'll leave a note explaining this.

gruberb reviewed (commented):
Looks good from the `iOS` side of things.

bendk commented:
Removed the `do-not-merge` label.  I think this could be ready to use now that UniFFI 0.28.2 is out.  Updated the PR slightly based on the new Swift bindgen features.
